### PR TITLE
Create chatrooms; post and delete messages. 

### PR DESCRIPTION
**Pull request by TeoReu**

### CODE REVIEW

File: server/routes/chat.go [hunk 1]
```diff
@@ -149,3 +143,166 @@ func userInList(a string, list []string) bool {
...
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	requestedUserID := string(bodyBytes)
+	id := userID + " " + requestedUserID
+	participants := []string{userID, requestedUserID}
+	cr := models.ChatRoom{ id, participants, nil}
+	
+	chatroomKey := datastore.NameKey("ChatRoom", cr.ID, nil)
+	_, err1 := dsClient.Put(ctx, chatroomKey, &cr)
```
Comment (chripell) [inline server/routes/chat.go:164]: it is more idiomatic to use _, err =

File: server/routes/chat.go [hunk 2]
```diff
@@ -149,3 +143,166 @@ func userInList(a string, list []string) bool {
...
+	if err1 != nil {
+		fmt.Println(err)
```
Comment (chripell) [inline server/routes/chat.go:166]: Is it OK to continue in case of error?

File: server/routes/chat.go [hunk 3]
```diff
@@ -149,3 +143,166 @@ func userInList(a string, list []string) bool {
...
+	}
+	
+	//add chatroomKey to CurrentUser.ChatRooms
+	k1 := datastore.NameKey("User", userID, nil)
+	
+	var currentUser models.User
+	if err := dsClient.Get(ctx, k1, &currentUser); err != nil {
+		log.Printf("Cannot retrieve user from DataStore: %v", err)
```
Comment (chripell) [inline server/routes/chat.go:174]: I would suggest to return an error to the client as well. You can user defer if you prefer not to do the same code all the time, or simply a helper function.

File: server/routes/chat.go [hunk 4]
```diff
@@ -149,3 +143,166 @@ func userInList(a string, list []string) bool {
...
+		log.Printf("Cannot parse message to json body: %v", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	//Setting the key name to be "<time><userID>"
+	keyname := message.Posted.String() + message.SenderID
+	newKey := datastore.NameKey("Message", keyname, nil)
+	newKey, err = dsClient.Put(ctx, newKey, &message)
+	if err != nil {
+		fmt.Println(err)
```
Comment (chripell) [inline server/routes/chat.go:228]: Is it OK to just continue if there is an error?

**Review by chripell — APPROVED** [review]

File: frontend/app/src/app/_services/chatroom.service.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import { HttpClient, HttpHeaders } from '@angular/common/http';
+import { Chatroom } from '../_models/chatroom'
+import { Message } from '../_models/message';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ChatroomService {
+
+  private chatroomsUrl = 'http://localhost:8000/messages';
```
Comment (Omn1) [inline frontend/app/src/app/_services/chatroom.service.ts:13]: Don't use hardcoded address. You need to put `apiEndpoint: "127.0.0.1:8000"` into some config

File: frontend/app/src/app/_services/profile.service.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+@Injectable({
+  providedIn: 'root'
+})
+export class ProfileService {
+  routeURL: string ;
+
+  constructor(private httpClient: HttpClient) { }
+
+  public upload(id, formData) {
+    this.routeURL = "http://localhost:8000/user/"+id+"/profile-image";
```
Comment (Omn1) [inline frontend/app/src/app/_services/profile.service.ts:14]: And here

File: server/routes/chat.go [hunk 7]
```diff
@@ -149,3 +143,177 @@ func userInList(a string, list []string) bool {
...
+	userID := userService.GetCurrentUserID(ctx)
+
+	bodyBytes, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Printf("Cannot read request body: %v", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	requestedUserID := string(bodyBytes)
+	id := userID + " " + requestedUserID
```
Comment (Omn1) [inline server/routes/chat.go:159]: I'd stick with the autogenerated chatroom id (you see how image id's are generated for reference)

File: server/routes/chat.go [hunk 8]
```diff
@@ -149,3 +143,177 @@ func userInList(a string, list []string) bool {
...
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	
+	//add chatroomKey to CurrentUser.ChatRooms
+	k1 := datastore.NameKey("User", userID, nil)
+	
+	var currentUser models.User
+	if err := dsClient.Get(ctx, k1, &currentUser); err != nil {
+		log.Printf("Cannot update user: %v", err)
```
Comment (Omn1) [inline server/routes/chat.go:177]: Why are there two `log.Printf`-s inside this `if`? This applies to several `if`-s below as well

**Review by Omn1 — CHANGES_REQUESTED** [review]
When adding and deleting messages in a conversation, you should check that the user belongs to that conversation

**Pull request closed without merging**